File: engine/Client/World.cs
namespace Client;

/// <summary>
/// Hosts the systems of a world. Add a constructed system to start it,
/// remove it to shut it down. Obtain one via <see cref="ECS.GetWorld"/>.
/// </summary>
/// <example>
/// <code>
/// var movement = new MovementSystem(dep);
/// world.AddSystem(movement);
/// await ecs.WaitForShutdownAsync();
/// world.RemoveSystem(movement);
/// </code>
/// </example>
public sealed class World : IAsyncDisposable
{
    private sealed record Entry(SystemRunner Runner, CancellationTokenSource Cts, Task Task);

    private readonly ECS _ecs;
    private readonly Lock _gate = new();
    private readonly Dictionary<SystemBase, Entry> _systems = new();

    internal World(ECS ecs, string name)
    {
        _ecs = ecs;
        Name = name;
    }

    public string Name { get; }

    public string NatsUrl => _ecs.NatsUrl;

    /// <summary>
    /// Runs <paramref name="system"/>: invokes OnAdd, connects to the transport,
    /// and starts its tick loop in the background.
    /// </summary>
    public void AddSystem(SystemBase system)
    {
        ArgumentNullException.ThrowIfNull(system);

        var cts = new CancellationTokenSource();
        var runner = new SystemRunner(system, NatsUrl);

        lock (_gate)
        {
            if (_systems.ContainsKey(system))
                throw new InvalidOperationException($"System '{system.SystemName}' has already been added.");

            // OnAdd runs here so authoring errors surface from AddSystem itself.
            system.InvokeOnAdd();
            _systems[system] = new Entry(runner, cts, RunSystemAsync(system, runner, cts.Token));
        }
    }

    /// <summary>
    /// Stops <paramref name="system"/> without waiting for it to finish shutting down.
    /// </summary>
    public void RemoveSystem(SystemBase system)
    {
        _ = RemoveSystemAsync(system).ContinueWith(
            t => Console.Error.WriteLine($"[{system.SystemName}] Shutdown error: {t.Exception}"),
            TaskContinuationOptions.OnlyOnFaulted);
    }

    /// <summary>
    /// Stops <paramref name="system"/> and waits until it has unregistered and disposed.
    /// </summary>
    public async Task RemoveSystemAsync(SystemBase system)
    {
        ArgumentNullException.ThrowIfNull(system);

        Entry? entry;
        lock (_gate)
        {
            if (!_systems.Remove(system, out entry))
                return;
        }

        await StopAsync(entry);
    }

    /// <summary>
    /// Stops every system currently added to this world.
    /// </summary>
    public async Task ShutdownAsync()
    {
        Entry[] entries;
        lock (_gate)
        {
            entries = _systems.Values.ToArray();
            _systems.Clear();
        }

        await Task.WhenAll(entries.Select(StopAsync));
    }

    public ValueTask DisposeAsync() => new(ShutdownAsync());

    private static async Task StopAsync(Entry entry)
    {
        try
        {
            await entry.Cts.CancelAsync();
            await entry.Task;
        }
        catch (OperationCanceledException) { /* graceful shutdown */ }
        finally
        {
            entry.Cts.Dispose();
        }
    }

    private async Task RunSystemAsync(SystemBase system, SystemRunner runner, CancellationToken ct)
    {
        try
        {
            await runner.ConnectAsync(ct);
            Console.WriteLine($"[{system.SystemName}] Starting...");
            await runner.RunAsync(ct);
        }
        catch (OperationCanceledException) when (ct.IsCancellationRequested) { /* graceful shutdown */ }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"[{system.SystemName}] Faulted: {ex}");
            _ecs.SignalShutdown();
            throw;
        }
        finally
        {
            system.InvokeOnRemove();
            await runner.DisposeAsync();
        }
    }
}

File: clients/typescript/src/world.ts
import type { NatsConnection } from "@nats-io/nats-core";
import { CoordinatorClient } from "./coordinator.js";
import { EntityCommandBuffer } from "./commands.js";
import { SystemRunner } from "./runner.js";
import type { SystemBase } from "./system.js";

/**
 * Hosts the systems of a world. Add a constructed system to start it, remove it to shut
 * it down.
 */
export class World {
  private readonly running = new Map<SystemBase, { controller: AbortController; task: Promise<void> }>();
  private readonly coordinator: CoordinatorClient;

  /**
   * Structural changes made outside any system — seed data, fixtures, one-off edits.
   * Apply them with `flush`.
   */
  readonly commands = new EntityCommandBuffer();

  constructor(
    readonly name: string,
    private readonly nats: NatsConnection,
  ) {
    this.coordinator = new CoordinatorClient(nats, `world:${name}`);
  }

  /**
   * Applies everything buffered in `commands`, waiting for the coordinator first so the
   * commands are not published into the void.
   */
  async flush(): Promise<void> {
    if (!this.commands.hasPendingCommands) return;

    await this.coordinator.waitForCoordinator();
    await this.coordinator.submit(this.commands.drain(), this.commands.schemas);
  }

  /** Runs a system: invokes onAdd, then starts its tick loop in the background. */
  addSystem(system: SystemBase): void {
    if (this.running.has(system)) {
      throw new Error(`System '${system.systemName}' has already been added.`);
    }

    // onAdd runs here so authoring errors surface from addSystem itself.
    system.invokeOnAdd();

    const controller = new AbortController();
    const runner = new SystemRunner(system, this.nats);
    const task = runner
      .run(controller.signal)
      .catch((error: unknown) => {
        if (!controller.signal.aborted) {
          console.error(`[${system.systemName}] Faulted: ${String(error)}`);
          throw error;
        }
      })
      .finally(() => system.invokeOnRemove());

    this.running.set(system, { controller, task });
  }

  async removeSystem(system: SystemBase): Promise<void> {
    const entry = this.running.get(system);
    if (entry === undefined) return;

    this.running.delete(system);
    entry.controller.abort();
    await entry.task.catch(() => undefined);
  }

  async shutdown(): Promise<void> {
    await Promise.all([...this.running.keys()].map((s) => this.removeSystem(s)));
  }
}

/**
 * Entry point for the SDK. Owns nothing but the worlds; the caller owns the connection.
 */
export class Ecs {
  private readonly worlds = new Map<string, World>();

  constructor(readonly nats: NatsConnection) {}

  world(name = "default"): World {
    let world = this.worlds.get(name);
    if (world === undefined) {
      world = new World(name, this.nats);
      this.worlds.set(name, world);
    }
    return world;
  }

  async shutdown(): Promise<void> {
    await Promise.all([...this.worlds.values()].map((w) => w.shutdown()));
  }
}
